text/0000-input-macros.md

```diff
@@ -0,0 +1,484 @@
+where
+    R: BufRead,
+    T: FromStr,
+    T::Err: std::fmt::Display + std::fmt::Debug,
+{
+    if let Some(prompt_args) = prompt {
+        match print_style {
+            PrintStyle::Continue => {
+                // Use print! for no newline
+                print!("{}", prompt_args);
+            }
+            PrintStyle::NewLine => {
+                // Use println! for adding a newline
+                println!("{}", prompt_args);
+            }
+        }
+        // Always flush so the user sees the prompt immediately
+        io::stdout().flush().map_err(InputError::Io)?;
+    }
+
+    let mut input = String::new();
+    let bytes_read = reader.read_line(&mut input).map_err(InputError::Io)?;
+    
+    // If 0, that's EOF — return Eof error
+    if bytes_read == 0 {
+        return Err(InputError::Eof);
+    }
+
+    let trimmed = input.trim_end_matches(['\r', '\n'].as_ref());
+    trimmed.parse::<T>().map_err(InputError::Parse)
+}
+
+/// A unified error type indicating either an I/O error, a parse error, or EOF.
+#[derive(Debug, PartialEq)]
+pub enum InputError<E> {
+    /// An I/O error occurred (e.g., closed stdin).
+    Io(io::Error),
+    /// Failed to parse the input into the desired type.
+    Parse(E),
+    /// EOF encountered (read_line returned 0).
+    Eof,
+}
+
+/// Defines how the prompt should be printed.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum PrintStyle {
+    /// Print the prompt without a trailing newline (uses `print!`).
+    Continue,
+    /// Print the prompt with a trailing newline (uses `println!`).
+    NewLine,
+}
+
+impl<E: std::fmt::Display + std::fmt::Debug> std::fmt::Display for InputError<E> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            InputError::Io(e) => write!(f, "I/O error: {}", e),
+            InputError::Parse(e) => write!(f, "Parse error: {}", e),
+            InputError::Eof => write!(f, "EOF encountered"),
+        }
+    }
+}
+
+impl<E: std::fmt::Display + std::fmt::Debug> std::error::Error for InputError<E> {}
+```
+
+Another thing to consider is that the `input!` and `inputln!` macros just 
+delegate to the `read_input_from` function, which is the core of the
+implementation. This function is generic over the reader type and the type
+that we want to parse. Besides that, we have a `PrintStyle` enum that is used to
+determine how to print the prompt. The `PrintStyle` enum has two variants:
+- `PrintStyle::Continue` that is used to print the prompt without a trailing 
+  newline.
+- `PrintStyle::NewLine` that is used to print the prompt with a trailing
+  newline.
+It allow us to use the same function for both macros and to have a more
+ergonomic way to print the prompt.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* Can lead to unnecessary buffer allocations in Rust programs when developers
+  don't realize that they could reuse a buffer instead. This could potentially
+  be remedied by a new Clippy lint.
```

Comment:
This implementation could potentially use a thread local String, or even a global `Mutex<String>` as a re-useable buffer to avoid allocating a `String` for every call. 

However, note that if `T` is `String` the call to `parse` will still allocate a new `String` (the currently implementation actually allocates two `Strings` in this case, since the `parse` call will make a copy of the original string.

Reply:
Do you have a different implementation in mind, or could you provide an example of your idea? It's just to have a clearer understanding of how it would work.